RemoteImageBufferProxy waits flushes in a dedicated work queue

kkinnunen-apple opened the pull request:
#### 2fed4020085ce890460e9adaa800e7c01afa01b2
<pre>
RemoteImageBufferProxy waits flushes in a dedicated work queue
<a href="https://bugs.webkit.org/show_bug.cgi?id=256694">https://bugs.webkit.org/show_bug.cgi?id=256694</a>
rdar://109257081

Reviewed by Matt Woodrow.

The wait semaphore would be moved to dedicated work queue and waited
upon. When the semaphore would signal, the work queue would wake up
the possible multiple waiters. This behavior is a bit redundant.

Instead, first waiter waits on the semaphore, and subsequent waiters
query the result of the first waiter. The timeout is not exact, as
it will be at maximum the timeout of the first waiter. However, all
current waiters have the same timeout. Subsequent commits might make
this more rigorous, when the pattern will be moved to new abstraction
IPC::Fence.

Removes DisplayListRecorderFlushIdentifier, it is not useful or used.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/DisplayListRecorderFlushIdentifier.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxyFlushFence::create):
(WebKit::RemoteImageBufferProxyFlushFence::~RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferProxyFlushFence::forceSignal):
(WebKit::RemoteImageBufferProxyFlushFence::RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxyFlushFence::WTF_GUARDED_BY_LOCK):
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::backingStoreWillChange):
(WebKit::RemoteImageBufferProxy::clearBackend):
(WebKit::RemoteImageBufferProxy::flushDrawingContext):
(WebKit::RemoteImageBufferProxy::flushDrawingContextAsync):
(WebKit::RemoteImageBufferProxy::createFlusher):
(WebKit::RemoteImageBufferProxy::prepareForBackingStoreChange):
(WebKit::RemoteImageBufferProxy::hasPendingFlush const): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::waitForDidFlush): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::markCompletedFlush): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::identifierForCompletedFlush const): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::cancel): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::lastSentFlushIdentifier const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::addPendingFlush): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:

Canonical link: <a href="https://commits.webkit.org/264240@main">https://commits.webkit.org/264240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/291986aa09b9d69ad581c84f168fe3e45be7c9ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6856 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7269 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8454 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7104 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7025 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10002 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8557 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5008 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14011 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6714 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6304 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9093 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6788 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5554 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6164 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1687 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6542 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -41,32 +41,78 @@
+        tracePoint(FlushRemoteImageBufferStart, reinterpret_cast<uintptr_t>(this));
+    }
+    Lock m_lock;
+    Condition m_condition;
```
This CV is unused, right?

kkinnunen-apple replied:
Yes. Removed.

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -41,32 +41,78 @@
     void flush() final
     {
-        m_flushState->waitForDidFlush(m_targetFlushIdentifier, Seconds::infinity());
+        m_flushState->waitFor(RemoteDisplayListRecorderProxy::defaultSendTimeout);
```
Previously we never timed out for flushing on a background thread (for rendering updates), and now we're adding a 3second timeout there.

I'm not super sure why we have different behaviour between threads currently, but it seems like changes to it should be done explicitly.

Assuming you think this is the right behaviour change, can you please add it to the commit message (or possibly even just land a separate one line commit, so this refactor becomes a no-op in terms of behaviour).

kkinnunen-apple replied:
#13873

mattwoodrow commented on Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
```diff
@@ -41,32 +41,77 @@
+        return m_signaled;
+    }
+
+    void forceSignal()
```
Is there any value in this function?

If a background thread is already waiting, then it'll block until that thread releases the lock (3 seconds at most), at which point we no longer need to signal the semaphore.

Could we just not bother? The only caller immediately drops its ref to this object afterwards.

I guess there's a chance that the background thread will start waiting after this call, but that seems like a race, where the current behaviour is we choose to block the main thread unnecessarily, and the alternative is to block the background thread unnecessarily.

kkinnunen-apple committed:
RemoteImageBufferProxy waits flushes in a dedicated work queue
https://bugs.webkit.org/show_bug.cgi?id=256694
rdar://109257081

Reviewed by Matt Woodrow.

The wait semaphore would be moved to dedicated work queue and waited
upon. When the semaphore would signal, the work queue would wake up
the possible multiple waiters. This behavior is a bit redundant.

Instead, first waiter waits on the semaphore, and subsequent waiters
query the result of the first waiter. The timeout is not exact, as
it will be at maximum the timeout of the first waiter. However, all
current waiters have the same timeout. Subsequent commits might make
this more rigorous, when the pattern will be moved to new abstraction
IPC::Fence.

Removes DisplayListRecorderFlushIdentifier, it is not useful or used.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Resources/Signposts/SystemTracePoints.plist:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/DisplayListRecorderFlushIdentifier.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxyFlushFence::create):
(WebKit::RemoteImageBufferProxyFlushFence::~RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferProxyFlushFence::forceSignal):
(WebKit::RemoteImageBufferProxyFlushFence::RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxyFlushFence::WTF_GUARDED_BY_LOCK):
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
(WebKit::RemoteImageBufferProxy::backingStoreWillChange):
(WebKit::RemoteImageBufferProxy::clearBackend):
(WebKit::RemoteImageBufferProxy::flushDrawingContext):
(WebKit::RemoteImageBufferProxy::flushDrawingContextAsync):
(WebKit::RemoteImageBufferProxy::createFlusher):
(WebKit::RemoteImageBufferProxy::prepareForBackingStoreChange):
(WebKit::RemoteImageBufferProxy::hasPendingFlush const): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::waitForDidFlush): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::markCompletedFlush): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::identifierForCompletedFlush const): Deleted.
(WebKit::RemoteImageBufferProxyFlushState::cancel): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::lastSentFlushIdentifier const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::addPendingFlush): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:

Canonical link: https://commits.webkit.org/264240@main
